### ISSUE TITLE
session.run() fails with Windows builtin CLI commands

### ISSUE THREAD

**Issue by schuderer** [theacodes/nox#220]
**Describe the bug**
TL;DR: Windows builtin cmd.exe commands like `del`, `mkdir`, etc. cannot be invoked using `session.run()` because they don't have their own binary file with a location in the filesystem, which `nox` requires.

I am trying to use noxfiles cross-platform (Mac/Linux/Win), using a conditional like the following:
```python
from sys import platform
import nox
@nox.session
def my_session(session):
    if platform == "win32":
        session.run("del", "docs\\mymodule.rst")
    else:
        session.run("rm", "-f", "docs/mymodule.rst")
    ...
```

This fails on Windows 10 with the message "Program del not found." at [nox.command.which()](https://github.com/theacodes/nox/blob/2fbced3f0766e8b29f455645c2c4083f30aa582c/nox/command.py#L32-L48)

If I replace the line
```python
        session.run("del", "docs\\mymodule.rst")
```
with
```python
        session.run("ping", "localhost", "-n", "1")
```
then it works.

Why the difference? `del` is a builtin `cmd.exe` command, while `ping.exe` is a binary which actually *has* a location. This can be checked on the command line: `where del` gives an error, while `where ping` prints something like `C:\windows\system32\ping.exe`.

**Expected behavior**
`session.run()` should be able to run common Windows CLI commands like `del`, `rmdir` etc.

**Current Workaround**
I'd probably create a batch file as a wrapper for Windows commands. I'd prefer to avoid this, because I'm trying to use `nox` to reduce the number of test/build/dist-related scripts.

**Suggestion**
`session.run()` uses `nox.command.which()` to find the full path to the binary, and `which()` raises an error if a path cannot be determined. I would suggest adding another `if` statement to `which()` testing whether the command is among the known builtin commands, akin to:
```python
from sys import platform
...
def which(program, path):
    """Finds the full path to an executable."""
    full_path = None
    ...

    if full_path:
        return full_path.strpath

    if platform == "win32" and program in WIN32_CMD_BUILDINS:
        return program

    logger.error("Program {} not found.".format(program))
    raise CommandFailed("Program {} not found".format(program))
```
where `WIN32_CMD_BUILDINS = ["ASSOC", "BREAK", "CALL" , "CD", "CHDIR", "CLS", "COLOR", "COPY", "DATE", "DEL", "DIR", "DPATH", "ECHO", "ENDLOCAL", "ERASE", "EXIT", "FOR", "FTYPE", "GOTO", "IF", "KEYS", "MD", "MKDIR", "MKLINK", "MOVE", "PATH", "PAUSE", "POPD", "PROMPT", "PUSHD", "REM", "REN", "RENAME", "RD", "RMDIR", "SET", "SETLOCAL", "SHIFT", "START", "TIME", "TITLE", "TYPE", "VER", "VERIFY", "VOL"]`

I did not find a way to easily programmatically determine the builtins, so they'd probably have to go to the top of the module as a constant. Provided there is a consensus on the validity of this bug report, and this is seen as an acceptable solution, I'd be happy to create a pull request.

**Comment by tswast** [comment]
FWIW, there are plenty of built-ins for bash shell on Linux, too.

https://www.gnu.org/software/bash/manual/html_node/Bourne-Shell-Builtins.html#Bourne-Shell-Builtins

I vote to keep `run` for running processes, since even on Windows there are multiple shell choices (bash via Linux subsystem, cmd, and PowerShell).

**Comment by tswast** [comment]
For `del`, I'd recommend the [os.remove](https://docs.python.org/3/library/os.html#os.remove) or [shutil.rmtree](https://docs.python.org/3/library/shutil.html#shutil.rmtree) functions.

**Comment by schuderer** [comment]
> FWIW, there are plenty of built-ins for bash shell on Linux, too.

Good point, now I get why we have `session.chdir()`! 😄 

> I vote to keep run for running processes [...]

Also an important distinction, thanks! I wasn't aware of `run()` being for processes only. When the [docs](https://nox.thea.codes/en/stable/config.html#nox.sessions.Session.run) said "Run a command",  I interpreted `run()` as a generic "run CLI stuff" method.

I'l try to list the options I see:
1. No changes to code nor docs (but I think the asymmetry of `rm` working here and `del` not working there will trip up some more users moving forward)
2. Make it clearer in docs what kinds of commands can and cannot be run, and point users to python commands like os.remove to help them along (ideally with example(s)). The burden to deal with asymmetries is with the user, but we help them along. Thanks @tswast for the pointers to those!
3. Create one or a few more session.<command> methods for the most pressing cases in analogy with `session.chdir()` (maybe in conjuction with the previous point
4. Create a generic method like `session.command()` to exist next to `session.run()` for running builtin commands (versus processes). Not sure if/how achievable to automatically determine which is which. Maybe if it's not found, it's automatically a command (or we error out if not)?
4. Relax the checks in `session.run()` and below to also work with builtin commands (where @tswast already said that this is probably not a good idea. If (!) we check all of them explicitly, it could become a burden to maintain the lists. Though here, too, we could just say: try to find the binary first, just run it if no binary was found (simply remove or catch the error in `command.which()`).

**Comment by tswast** [comment]
My personal favorite of these is (2). I think some nox could use some how-to guides with concrete use cases, maybe even comparing with other common options for managing test sessions (such as batch and shell scripts).

**Comment by schuderer** [comment]
Epilogue (non-reopening): The closing documentation PR is providing a minimum-viable solution/workaround to this issue. However, documenting the ins and outs of multi-platform nox files (with the tips given here, particularly by @tswast) might still be worthwhile, though probably for the tutorial instead of the api doc. Potentially in conjunction with #149 (write a guide on nox with CI).
